src/main/java/com/metamx/common/parsers/JSONPathParser.java

```diff
@@ -0,0 +1,177 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class JSONPathParser implements Parser<String, Object>
+{
+  private Map<String, Pair<FieldType, JsonPath>> fieldPathMap;
+  private List<FieldSpec> fieldSpecs;
+  private final boolean useFieldDiscovery;
+
+  public JSONPathParser(List<FieldSpec> fieldSpecs, boolean useFieldDiscovery)
+  {
+    this.fieldSpecs = fieldSpecs;
+    this.fieldPathMap = generateFieldPaths(fieldSpecs);
+    this.useFieldDiscovery = useFieldDiscovery;
+    setJacksonProvider();
+  }
+
+  @Override
+  public List<String> getFieldNames()
+  {
+    return null;
+  }
+
+  @Override
+  public void setFieldNames(Iterable<String> fieldNames)
+  {
+  }
+
+  @Override
+  public Map<String, Object> parse(String input)
+  {
+    Map<String, Object> map = new LinkedHashMap<>();
+    Map<String, Object> document = (Map<String, Object>) Configuration.defaultConfiguration().jsonProvider().parse(input);
+    for (Map.Entry<String, Pair<FieldType, JsonPath>> entry : fieldPathMap.entrySet()) {
+      String fieldName = entry.getKey();
+      Pair<FieldType, JsonPath> pair = entry.getValue();
+      JsonPath path = pair.rhs;
+      Object parsedVal;
+      if (pair.lhs == FieldType.ROOT) {
+        parsedVal = document.get(fieldName);
+      } else {
+        parsedVal = path.read(document);
+      }
+      map.put(fieldName, parsedVal);
+    }
+    if(useFieldDiscovery) {
+      discoverFields(map, document);
+    }
+    return map;
+  }
+
+  private Map<String, Pair<FieldType, JsonPath>> generateFieldPaths(List<FieldSpec> fieldSpecs)
+  {
+    Map<String, Pair<FieldType, JsonPath>> map = new LinkedHashMap<>();
+    for (FieldSpec fieldSpec : fieldSpecs) {
+      String fieldName = fieldSpec.getName();
+      JsonPath path = JsonPath.compile(fieldSpec.getExpr());
+      Pair<FieldType, JsonPath> pair = new Pair<>(fieldSpec.getType(), path);
+      map.put(fieldName, pair);
+    }
+    return map;
+  }
+
+  private void setJacksonProvider()
+  {
+    Configuration.setDefaults(
+        new Configuration.Defaults()
+        {
+
+          private final JsonProvider jsonProvider = new JacksonJsonProvider();
+          private final MappingProvider mappingProvider = new JacksonMappingProvider();
+
+          @Override
+          public JsonProvider jsonProvider()
+          {
+            return jsonProvider;
+          }
+
+          @Override
+          public MappingProvider mappingProvider()
+          {
+            return mappingProvider;
+          }
+
+          @Override
+          public Set<Option> options()
+          {
+            return EnumSet.noneOf(Option.class);
+          }
+        }
+    );
+  }
+
+  private void discoverFields(Map<String, Object> map, Map<String, Object> document) {
+    for(String field : document.keySet()) {
+      if(!map.containsKey(field)) {
+        Object val = document.get(field);
+        if(!(val instanceof Map) && !(val instanceof List)) {
```

Comment:
We still want to include Lists of basic types (like Strings or numbers) as those are valid root level things.